packages/dynamic-import/server.js

```diff
@@ -1,48 +1,129 @@
+    // abandon the rest of the logic in this module.
+    //
+    // If api.use("webapp") appeared in dynamic-import/package.js, then
+    // Package.webapp would always be defined here, of course, but that
+    // would be a bad idea, because the dynamic-import package should not
+    // single-handedly force a dependency on webapp if the program does
+    // not otherwise need a web server (e.g., when the program is an
+    // isopacket or build plugin instead of a web application).
+    //
+    // Note that the client.js module (imported above) still defines
+    // Module.prototype.dynamicImport, which will work as long as no
+    // modules need to be fetched.
+    return;
+  }
 
-Object.keys(dynamicImportInfo).forEach(platform => {
+  Object.keys(dynamicImportInfo).forEach(setUpPlatform);
+
+  Package.webapp.WebApp.connectHandlers.use(
+    "/__dynamicImport",
+    middleware
+  );
+});
+
+function setUpPlatform(platform) {
   const info = dynamicImportInfo[platform];
+
   if (info.dynamicRoot) {
     info.dynamicRoot = pathNormalize(info.dynamicRoot);
   }
-});
 
-Meteor.methods({
-  __dynamicImport(tree) {
-    check(tree, Object);
-    this.unblock();
-
-    const platform = this.connection ? "web.browser" : "server";
-    const pathParts = [];
-
-    function walk(node) {
-      if (node && typeof node === "object") {
-        Object.keys(node).forEach(name => {
-          pathParts.push(name);
-          node[name] = walk(node[name]);
-          assert.strictEqual(pathParts.pop(), name);
-        });
-      } else {
-        return read(pathParts, platform);
+  if (platform === "server") {
+    client.setSecretKey(info.key = randomId(40));
+  }
+}
+
+function randomId(n) {
+  let s = "";
+  while (s.length < n) {
+    s += Math.random().toString(36).slice(2);
+  }
+  return s.slice(0, n);
+}
+
+function middleware(request, response) {
+  assert.strictEqual(request.method, "POST");
+  const chunks = [];
+  request.on("data", chunk => chunks.push(chunk));
+  request.on("end", () => {
+    response.setHeader("Content-Type", "application/json");
+    response.end(JSON.stringify(readTree(
+      JSON.parse(Buffer.concat(chunks)),
+      getPlatform(request)
+    )));
+  });
+}
+
+function getPlatform(request) {
+  let platform = "web.browser";
+
+  // If the __dynamicImport request includes a secret key, and it matches
+  // dynamicImportInfo[platform].key, use platform instead of the default
+  // platform, web.browser.
+  const secretKey = request.query.key;
+
+  if (typeof secretKey === "string") {
+    Object.keys(dynamicImportInfo).some(p => {
+      if (secretKey === dynamicImportInfo[p].key) {
+        platform = p;
+        return true;
       }
-      return node;
-    }
+    });
+  }
+
+  return platform;
+}
+
+function readTree(tree, platform) {
+  const pathParts = [];
 
-    return walk(tree);
+  function walk(node) {
+    if (node && typeof node === "object") {
+      let empty = true;
```

Comment:
The number of lines inside this `if` conditional has grown a bit with this commit.  It could improve readability, and get rid of the cliff if the `else` and its `return` were first:

```js
function walk(node) {
  if (typeof node !== "object") {
    return read(pathParts, platform);
  }

  let empty = true;
  // ... so on.
  return node;
}
```